runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -0,0 +1,389 @@
+local query = require'vim.treesitter.query'
+local language = require'vim.treesitter.language'
+
+local LanguageTree = {}
+LanguageTree.__index = LanguageTree
+
+-- Represents a single treesitter parser for a language.
+-- The language can contain child languages with in it's range,
+-- hence the tree.
+--
+-- @param source Can be a bufnr or a string of text to parse
+-- @param lang The language this tree represents
+-- @param opts Options table
+-- @param opts.queries A table of language to injection query strings
+--                     This is useful for overridding the built in runtime file
+--                     searching for the injection language query per language.
+function LanguageTree.new(source, lang, opts)
+  language.require_language(lang)
+  opts = opts or {}
+
+  local custom_queries = opts.queries or {}
+  local self = setmetatable({
+    _source=source,
+    _lang=lang,
+    _children = {},
+    _regions = {},
+    _trees = {},
+    _opts = opts,
+    _injection_query = custom_queries[lang]
+      and query.parse_query(lang, custom_queries[lang])
+      or query.get_query(lang, "injections"),
+    _valid = false,
+    _parser = vim._create_ts_parser(lang),
+    _callbacks = {
+      changedtree = {},
+      bytes = {},
+      child_added = {},
+      child_removed = {}
+    },
+  }, LanguageTree)
+
+
+  return self
+end
+
+-- Invalidates this parser and all it's children
+function LanguageTree:invalidate()
+  self._valid = false
+
+  for _, child in ipairs(self._children) do
+    child:invalidate()
+  end
+end
+
+-- Returns all trees this language tree contains.
+-- Does not include child languages.
+function LanguageTree:trees()
+  return self._trees
+end
+
+-- Gets the language of this tree layer.
+function LanguageTree:lang()
+  return self._lang
+end
+
+-- Determines whether this tree is valid.
+-- If the tree is invalid, `parse()` must be called
+-- to get the an updated tree.
+function LanguageTree:is_valid()
+  return self._valid
+end
+
+-- Returns a map of language to child tree.
+function LanguageTree:children()
+  return self._children
+end
+
+-- Parses all defined regions using a treesitter parser
+-- for the language this tree represents.
+-- This will run the injection query for this language to
+-- determine if any child languages should be created.
+function LanguageTree:parse()
+  if self._valid then
+    return self._trees
+  end
+
+  local parser = self._parser
+  local changes = {}
+
+  self._trees = {}
+
+  -- If there are no ranges, set to an empty list
+  -- so the included ranges in the parser ar cleared.
+  if self._regions and #self._regions > 0 then
+    for _, ranges in ipairs(self._regions) do
+      parser:set_included_ranges(ranges)
+
+      local tree, tree_changes = parser:parse(nil, self._source)
+
+      table.insert(self._trees, tree)
+      vim.list_extend(changes, tree_changes)
+    end
+  else
+    local tree, tree_changes = parser:parse(nil, self._source)
```

Comment:
Here, this disabled incremental parsing, we need to have an association between trees and regions, to be able to edit the trees, and parse them incrementaly.

Reply:
Should be good now, but that caused test failures, resolving for now.